Re-add non-boilerplate code and add tests

seanh opened the pull request:
This re-adds the non-boilerplate files that https://github.com/hypothesis/pip-sync-faster/pull/5 (which this PR is based on) removed in order to facilitate code review. Also adds tests.

Once again a review of the meat is required (`README.md`, `src/*` and `tests/*`) and the cookiecutter stuff (`.cookiecutter/*`) can be reviewed when we review the cookiecutter itself.

# How to test manually

The pip-sync-faster concept is pretty simple: it calls `pip-sync` and stashes hashes of the requirements files in a `pip_sync_faster.json` file within the venv. If you run it again with the same requirements files and you haven't changed any of the requirements files then it won't call `pip-sync` again. If any of the requirements files has changed or if you call it with a different set of requirements files (even a subset of the ones you called it with last time) then it will call `pip-sync` again.

The CLI of `pip-sync-faster` is the same as that of `pip-sync` and it passes all command line options and arguments blindly to `pip-sync`. It also echoes `pip-sync`'s exit status.

To test it:

* Create and activate a venv

* Create some test requirements files

* Install `pip-sync-faster` in the venv:

  ```terminal
  pip install -e /path/to/pip-sync-faster
  ```

* It should be pretty obvious to see that it's running pip-sync:

  ```terminal
  pip-sync-faster requirements.txt
  ```

* There is one testing gotcha: `pip-sync-faster` must itself be in the requirements file, otherwise it'll uninstall itself (it'll call `pip-sync requirements.txt` which'll uninstall anything that's not in `requirements.txt`, including `pip-sync-faster`).

  In testing this is a double-gotcha because even if `requirements.txt` contains `pip-sync-faster` it will replace the local development version with a copy from PyPI.

  So in testing a local development version you have to reinstall it every time:

  ```terminal
  pip install -e /path/to/pip-sync-faster
  ```

* If you run `pip-sync-faster` again with the same requirements file it won't call `pip-sync`:

  ```terminal
  pip-sync-faster requirements.txt
  ```

Other things to test:

* If you modify the requirements file and then re-call `pip-sync-faster` it will call `pip-sync`

* If you call it with a different requirements file it will call `pip-sync`. If you then call it with the first requirements file again it'll call `pip-sync` again

* Calling it with multiple requirements files. If you call it again without changing any of the requirements files it won't call `pip-sync`. If you change any of them it will

* If you call it with a different set of requirements files (even a subset of what you previously called it with) it will call `pip-sync`

seanh committed:
Re-add non-boilerplate code and tests

seanh commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,181 @@
+        return mocker.patch("pip_sync_faster.main.pip_sync_maybe", autospec=True)
+
+
+class File:
+    """Metadata about a file."""
+
+    def __init__(self, relpath):
+        self.relpath = relpath
+        self.abspath = abspath(relpath)
+
+        hashobj = hashlib.sha512()
+        with open(self.abspath, "rb") as file:
+            hashobj.update(file.read())
+
+        self.hash = hashobj.hexdigest()
```
Maybe the tests should just use the `SrcFile` class from `main.py`: it's exactly the same thing as this class.

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
-    return "Hello, world!"
+def get_hash(path):
+    """Return the hash of the given file."""
+    hashobj = hashlib.sha512()
 
+    with open(path, "rb") as file:
+        hashobj.update(file.read())
+
+    return hashobj.hexdigest()
```
```suggestion
    digest = hashlib.sha512()

    with open(path, "rb") as handle:
        digest.update(handle.read())

    return digest.hexdigest()
```

seanh replied:
The object that `hashlib.sha512()` returns isn't called a "digest" (that's the name for what methods like `hashobj.digest()` and `hashobj.hexdigest()` return). Unfortunately the examples in the `hashlib` docs all use single-letter variable names for these objects and the letter varies from example to example: `m` or `h`. But the text of the `hashlib` docs refers to them as "hash objects"

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+
+    try:
+        with open(cached_hashes_path, "r", encoding="utf-8") as cached_hashes_file:
+            cached_hashes = json.load(cached_hashes_file)
```
```suggestion
        with open(cached_hashes_path, "r", encoding="utf-8") as handle:
            cached_hashes = json.load(handle)
```

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+    return {abspath(path): get_hash(abspath(path)) for path in paths}
+
+
+def pip_sync_maybe(src_files, args=None):
```
```suggestion
def pip_sync(src_files, args=None):
```

I don't think we need to add `_maybe` on all functions with an early return, raises an exception, could ever go wrong etc. If this is an important detail then a doc-string is probably the place.

seanh replied:
I've renamed this to `pip_sync_faster()` instead

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
     parser = ArgumentParser()
-    parser.add_argument("-v", "--version", action="store_true")
+    parser.add_argument("--version", action="store_true")
+    parser.add_argument("src_files", nargs="*")
```
Adding help strings here is probably the best docs we can add to a command line script. 

Normally we place the arg parser config at the top of the file and use the parser object here. This is a nice way to tip off the reader this is to be used as a script, not a library.

seanh replied:
I've added docstrings based on `pip-sync`'s ones

seanh replied:
The argparse stuff ended up getting moved into `cli.py` so where have a clear and obvious place to find the CLI code

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+    args = parser.parse_known_args(argv)
 
-    if args.version:
-        print(version("pip-sync-faster"))
```
So this is trying to get around the fact we'd like to accept a lot of args that `pip-sync` takes without having to list them all?

seanh replied:
This just prints the version number

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+    try:
+        pip_sync_maybe(args[0].src_files, argv[1:])
+    except CalledProcessError as err:
+        sys.exit(err.returncode)
```
```suggestion
        sys.exit(err.returncode)
        
if __name__ == '__main__':
    entry_point()
```

This will allow us to run this script directly without fuss.

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+import hashlib
```
```suggestion
#!/usr/bin/env python
import hashlib
```

Along with the `__main__` section at the bottom this would allow you to call this directly on the command line which might be handy for testing.

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+        return
+
+    # The hashes did not match the cached ones. This can happen if:
+    #
```
```suggestion
    #
    # * This is the first time we've been called
```
?

seanh replied:
Added this

jon-betts commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,156 @@
+        return [dev_txt_path, format_txt_path]
+
+
+class TestEntryPoint:
```
In many cases I think this is something you _could_ skip, but I think testing it in this case is actually a good idea as we're doing slightly fancy arg merging stuff.

jon-betts commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,156 @@
+        # removed from the venv.
+        run.assert_called_once_with(["pip-sync", *requirements_files[1:]], check=True)
+        # It should update the cache to contain only format.txt.
+        assert_hashes_cached(requirements_files[1:])
```
Are the three tests cases above really different? Might be an opportunity for less here.

jon-betts commented on README.md:
````diff
@@ -13,3 +13,45 @@ For installation instructions see [INSTALL.md](https://github.com/hypothesis/pip
+sys     0m0.008s
+```
+
+pip-sync-faster does this by saving hashes of the given requirements files in a
````
```suggestion
`pip-sync-faster` does this by saving hashes of the given requirements files in a
```
? Might stop the names from blending in quite so much?

seanh replied:
:+1:

jon-betts commented on README.md:
```diff
@@ -13,3 +13,45 @@ For installation instructions see [INSTALL.md](https://github.com/hypothesis/pip
+requirements files, but calling pip-sync-faster won't.
+
+If you can live with this limitation then you can use pip-sync-faster and save
+yourself a few hundred milliseconds.  If not you should just use pip-sync.
```
```suggestion
yourself a few hundred milliseconds. If not you should just use pip-sync.
```

We could also mention that:

 * It's very useful for debugging
 * If you want to undo it you can delete the env and it will get recreated

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+
+    # The hashes did not match the cached ones. This can happen if:
+    #
+    # * One or more of the requirements files has changed: its current hash
+    #   differs from its cached one
+    # * pip-sync-faster was called with a different set of requirements files
+    #   than last time: either one or more of the requirements files are
+    #   missing from the cached hashes, or the cached hashes contain one or
+    #   more requirements files that are missing from the command line args
```
```suggestion
    # * One or more of the requirements files has changed
    # * We were called with a different set of requirements files
```
Some redundancy here?

seanh replied:
:+1: Done

jon-betts commented on src/pip_sync_faster/main.py:
```diff
@@ -1,18 +1,75 @@
+    # Replace the cached hashes file with one containing the correct hashes for
+    # the requirements files that pip-sync-faster was called with this time.
+    with open(cached_hashes_path, "w", encoding="utf-8") as cached_hashes_file:
+        json.dump(hashes, cached_hashes_file)
```
```suggestion
    with open(cached_hashes_path, "w", encoding="utf-8") as handle:
        json.dump(hashes, handle)
```

seanh replied:
:+1:

jon-betts commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,156 @@
+        self, cache_hashes, run, requirements_files
+    ):
+        # Make sure that all the cached hashes are present and matching.
+        cache_hashes(main.get_hashes(requirements_files))
```
There's something a bit iffy about using `get_hashes` to test this which is a private function of the method under test?

Some pre-baked values might be a better control, and probably simpler as well, as we can just have a couple of fixture files (fixture in the regular testing meaning of the word, rather than `pytest`).

seanh replied:
:+1: Replaced this with pre-baked hashes

jon-betts commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,156 @@
+            capsys.readouterr().out.strip()
+            == f"pip-sync-faster, version {version('pip-sync-faster')}"
+        )
+        assert exc_info.value.code is None
```
Do we want / need a `version` option? Would usual pip type versions of the installed package do?

seanh replied:
Yeah, I think we want to have a `--version` (for `pip-sync-faster` and for our CLI packages in general). It's generated for us by the cookiecutter so it's no effort, and it provides an easy way to ask someone what version they have. The `version()` function works with setuptools_scm, see https://www.seanh.cc/2022/05/21/publishing-python-packages-from-github-actions/#the---version-command

jon-betts commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,156 @@
+            # There isn't actually a virtualenv here but it doesn't matter:
+            # this is just where pip-sync-faster will look for and create
+            # its cached hashes file.
+            "VIRTUAL_ENV": str(tmp_path.absolute()),
```
:+1:

jon-betts commented on tests/unit/pip_sync_faster/main_test.py:
```diff
@@ -1,6 +1,156 @@
+
+@pytest.fixture(autouse=True)
+def run(mocker):
+    return mocker.patch("pip_sync_faster.main.run", autospec=True)
```
This mock is only used in `TestPipSyncMaybe` and could be moved to be local to it

seanh replied:
We _never_ want any tests to call the real `pip-sync` (including tests that're currently in the file and others that might be added in future) so it's best for this to be a module-level autospecced fixture. This is how we generally do it when patching imports

jon-betts reviewed (commented):
### Would this benefit from following a more standard script pattern?

I think there's an opportunity here to make it clearer that this is a script, which would help understanding how it works and also avoid all of the testing woes noted in the ticket.

We have a pretty common pattern in a lot of our scripts (along with `chmod +x` for extra points) which I think we can follow here:

```python
#!/usr/bin/env python

from argparse import ArgumentParser

parser = ArgumentParser(description="A tool for ...")
...

def main():
    args = parser.parse_args()
    ...

if __name__ == '__main__':
    main()
```

That won't work verbatim here as there are more complex arg parsing requirements, but the general structure of the script could be followed. This has a few nice features:

* It's obvious it's intended to be used as a script (not a library)
* The docs for how to use it are integrated and front and center
* You can use it directly on the command line, or test it

This would avoid having to install it into the env in order to test it.

### Other stuff

The rest is mostly just random naming suggestions. 

The only thing I'd say is definitely worth thinking about is whether pre-baked hashes in the tests is a better way to go. At the moment they use the code which is indirectly under test is used to setup the fixture.

Another approach might be to have tests which explicitly drive it from the outside, and call the method twice. Changing things inbetween or not. This way you'd only be testing the externally observable behavior without having to bake in any hashes if that made you unconfortable.

seanh commented on pyproject.toml:
```diff
@@ -43,6 +43,9 @@ ignore = [
 parallel = true
 source = ["pip_sync_faster", "tests/unit"]
+omit = [
+    "*/pip_sync_faster/__main__.py",
```
This'll go into the cookiecutter, `__main__.py` will always be omitted from coverage

seanh commented:
This is ready for re-review: 

* I've added a `__main__.py` so that the package can be executed directly with `python3 -m pip_sync_faster <args>`. Added manual testing instructions to `HACKING.md` that use this.
* Re-named `pip_sync_maybe()` to just `sync()`
* Added help strings to the CLI (see `tox -qe dev --run-command 'pip-sync-faster --help'`)
* The tests now use pre-baked hashes
* Accepted various naming suggestions (`handle` etc)
* Accepted various other minor suggestions

### File layout

I spent some time thinking about the best file layout for the cookiecutter to use for packages like this one that have a command line interface. This is what I came up with:

```
pip-sync-faster/
  setup.cfg     <- Declares a console_script that calls cli.py::cli()
  pip_sync_faster/
    __init__.py <- Empty (except maybe some imports)
    __main__.py <- Just imports and calls cli.py::cli()
    cli.py      <- Contains the argparse code and imports and calls something from core.py
    core.py     <- The actual code
  tests/
    unit/
      pip_sync_faster/
        cli_test.py
        core_test.py
```

If you install the package and run `pip-sync-faster` it goes `setup.cfg` -> `cli.py::cli()` -> `core.py`, and `__main__.py` isn't involved. If you run the package directly with `python3 -m pip_sync_faster` it's `__main__.py` -> `cli.py::cli()` -> `core.py` and `setup.cfg` isn't involved.

For packages that don't have a CLI the cookiecutter will use the same layout but with the CLI-related files missing:

```
pip-sync-faster/
  setup.cfg
  pip_sync_faster/
    __init__.py
    core.py
  tests/
    unit/
      pip_sync_faster/
        core_test.py
```

Here's why I went for this layout:

* Empty `__init__.py` file. I know that some people don't like code in `__init__.py` and people new to Python often find it surprising. I don't want to have to have a file named `__init___test.py`. I don't want to have to open `__init__.py` often because it plays badly with autocomplete (since there are lots of files with that name). Packages might want to put some imports in `__init__.py` in order to hoist them into the package's top-level namespace, the cookiecutter won't do any of these for you but won't delete any that you put there.

* Minimal `__main__.py` with no tests. This is what the Python docs says is [idiomatic usage of `__main__.py`](https://docs.python.org/3/library/__main__.html#id1): when Python developers see a `__main__.py` file their eyes can pass right over it because they can expect it won't contain any code except to import and call an entry point function. This is how the runnable stdlib packages do it. Also it's nice to avoid a `__main___test.py`.

* Separate `cli.py` file that contains the `argparse` code and imports and calls things from `core.py`. Reasons for separating `cli.py` out from `core.py`:
  * Two smaller modules instead of one bigger one. Not really necessary for a small package like `pip-sync-faster` but scales well as the code size grows
  * A file named `cli.py` makes it obvious where the CLI code is going to be found
  * You can cookiecutter a package _without_ a CLI then add one later. Change `console_script` to `"yes"` in `cookiecutter.json` and run `make template`. All the cookiecutter has to do is add the `console_script` to the templated `setup.cfg` file and add the `__main__.py`, `cli.py` and `cli_test.py` files. It doesn't need to modify `core.py` and `core_test.py` which it can't really touch because the user will have changed these files. If the CLI code were in `core.py` you couldn't really use `make template` do add a CLI later like this.

* I chose the name `core.py` because `main.py` (what it was previously called) would clash with `__main__.py`. I also didn't want to call it `pip_sync_faster.py` because this name would have to vary between projects and because it makes for awkward imports especially if the package wants to have a _function_ named `pip_sync_faster()` as the existence of a module with this same name blocks hoisting the function into `__init__.py`. You can end up with stuff like `from pip_sync_faster.pip_sync_faster import pip_sync_faster`.

  `core.py` is expected to be changed by the user including perhaps splitting it up into multiple files or just keeping it as a single file but renaming it. Here in `pip-sync-faster` I've renamed it to `sync.py` as that seems a more appropriate name specific to this project. In tox plugins it might get renamed to `plugin.py`, etc.

I like that we can change around the internals of the package without breaking its interface. For example we could change the contents of `cli.py` or rename that file etc and as long as we update `setup.cfg` and `__main__.py` the `pip-sync-faster` and `python3 -m pip_sync_faster` commands won't change.

seanh committed:
Don't call sys.exit in cli.py

It's enough for a console script entry point to just return an int and
setuptools makes sure that the command exits with that int as its
status.

This does mean that we have to add a sys.exit() to __main__.py so that
it mirrors the same behaviour.

seanh committed:
Add more exit_code test assertions

seanh committed:
Rename pip_sync_faster() to just sync()

seanh committed:
Rename core.py to sync.py

A more appropriate/specific name for this particular project.

seanh commented on setup.cfg:
```diff
@@ -17,13 +17,15 @@ package_dir =
 [options.entry_points]
 console_scripts =
-    pip-sync-faster = pip_sync_faster.main:entry_point
+    pip-sync-faster = pip_sync_faster.cli:cli
```
I'll update this in the cookiecutter: `main.py::entry_point()` renamed to `cli.py::cli()`

seanh commented on src/pip_sync_faster/__main__.py:
```diff
@@ -0,0 +1,5 @@
+import sys
+
+from pip_sync_faster.cli import cli
+
+sys.exit(cli())
```
It'll be a good idea to move this `__main__.py` into the cookiecutter where it can be implemented correctly once and for all. For one thing we can implement `__main__.py` in the idiomatic way (as here), and we can omit it from test coverage. Also it's actually not as trivial to implement as you might expect:

`cli()` is used as both the setuptools entry point function (specified in `setup.cfg`) and the top-level function for `__main__.py` to call. A setuptools entry point function is expected to return something appropriate for being passed to [`sys.exit()`](https://docs.python.org/3/library/sys.html#sys.exit): `None` or `0` to exit successfully, an int to exit with that error code, or any printable object (such as a string) that will be printed to stderr before exiting with `1`. You can see this by looking at the `pip-sync-faster` script that setuptools generates when you install the package:

```terminal
$ cat .tox/dev/bin/pip-sync-faster
#!/home/seanh/Projects/pip-sync-faster/.tox/dev/bin/python
# -*- coding: utf-8 -*-
import re
import sys
from pip_sync_faster.cli import cli
if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
    sys.exit(cli())
```

A `__main__.py` or an `if __name__ == "__main__"` block should do the same thing: `sys.exit(cli())`.

This means that `cli()` itself shouldn't call `sys.exit()`, it should just return `None` or `0` to exit successfully or `return 1` or `return "Some error message"` to exit with an error

marcospri replied:
TIL about sys.exit with a function, nice.

seanh committed:
Disable a PyLint warning

seanh commented on src/pip_sync_faster/cli.py:
```diff
@@ -0,0 +1,28 @@
+from pip_sync_faster.sync import sync
+
+
+def cli(_argv=None):  # pylint:disable=inconsistent-return-statements
```
PyLint will force you to put `return 0` or `return None` in every branch of the function (including at the very bottom) which just seems annoying for an entry point function.

seanh commented on src/pip_sync_faster/cli.py:
```diff
@@ -0,0 +1,28 @@
+        "src_files", nargs="*", help="the requirements.txt files to synchronize"
+    )
+
+    args = parser.parse_known_args(_argv)
```
`_argv` (which is used by the tests) defaults to `None` and `parse_known_args(None)` causes `argparse` to parse the args from `sys.argv`

seanh commented on src/pip_sync_faster/sync.py:
```diff
@@ -0,0 +1,51 @@
+    # Replace the cached hashes file with one containing the correct hashes for
+    # the requirements files that pip-sync-faster was called with this time.
+    with open(cached_hashes_path, "w", encoding="utf-8") as handle:
+        json.dump(hashes, handle)
```
When you see it split out into its own file like this the actual core logic of pip-sync-faster is very simple (and only about 40 lines of code without comments and docstrings)

seanh committed:
Tweak code comment

marcospri commented on src/pip_sync_faster/sync.py:
```diff
@@ -0,0 +1,50 @@
+
+
+def sync(src_files):
+    cached_hashes_path = Path(environ["VIRTUAL_ENV"]) / "pip_sync_faster.json"
```
I reckon `environ["VIRTUAL_ENV"]`  is the right place to store it. It would be just noise in the project's root.

marcospri commented on src/pip_sync_faster/sync.py:
```diff
@@ -0,0 +1,50 @@
+
+def get_hash(path):
+    """Return the hash of the given file."""
+    hashobj = hashlib.sha512()
```
You could probably squeeze a few microseconds switching to sha256, if git considers that future-proof should be enough for this.

It doesn't probably matter.

marcospri reviewed (approved):
This works as expected. It's easy to understand and comes with good docs :+1:
